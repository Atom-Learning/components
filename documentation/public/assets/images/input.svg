<svg width="293" height="40" viewBox="0 0 293 40" fill="none" xmlns="http://www.w3.org/2000/svg">
<rect x="0.5" y="0.5" width="292" height="39" rx="3.5" fill="white"/>
<path d="M12.7727 15.6136V14.3636H21.5V15.6136H17.8409V26H16.4318V15.6136H12.7727ZM25.8864 26.1818C25.0455 26.1818 24.3201 25.9962 23.7102 25.625C23.1042 25.25 22.6364 24.7273 22.3068 24.0568C21.9811 23.3826 21.8182 22.5985 21.8182 21.7045C21.8182 20.8106 21.9811 20.0227 22.3068 19.3409C22.6364 18.6553 23.0947 18.1212 23.6818 17.7386C24.2727 17.3523 24.9621 17.1591 25.75 17.1591C26.2045 17.1591 26.6534 17.2348 27.0966 17.3864C27.5398 17.5379 27.9432 17.7841 28.3068 18.125C28.6705 18.4621 28.9602 18.9091 29.1761 19.4659C29.392 20.0227 29.5 20.7083 29.5 21.5227V22.0909H22.7727V20.9318H28.1364C28.1364 20.4394 28.0379 20 27.8409 19.6136C27.6477 19.2273 27.3712 18.9223 27.0114 18.6989C26.6553 18.4754 26.2348 18.3636 25.75 18.3636C25.2159 18.3636 24.7538 18.4962 24.3636 18.7614C23.9773 19.0227 23.6799 19.3636 23.4716 19.7841C23.2633 20.2045 23.1591 20.6553 23.1591 21.1364V21.9091C23.1591 22.5682 23.2727 23.1269 23.5 23.5852C23.7311 24.0398 24.0511 24.3864 24.4602 24.625C24.8693 24.8598 25.3447 24.9773 25.8864 24.9773C26.2386 24.9773 26.5568 24.928 26.8409 24.8295C27.1288 24.7273 27.3769 24.5758 27.5852 24.375C27.7936 24.1705 27.9545 23.9167 28.0682 23.6136L29.3636 23.9773C29.2273 24.4167 28.9981 24.803 28.6761 25.1364C28.3542 25.4659 27.9564 25.7235 27.483 25.9091C27.0095 26.0909 26.4773 26.1818 25.8864 26.1818ZM32.3054 17.2727L34.3963 20.8409L36.4872 17.2727H38.0327L35.2145 21.6364L38.0327 26H36.4872L34.3963 22.6136L32.3054 26H30.7599L33.5327 21.6364L30.7599 17.2727H32.3054ZM43.7415 17.2727V18.4091H39.2188V17.2727H43.7415ZM40.5369 15.1818H41.8778V23.5C41.8778 23.8788 41.9328 24.1629 42.0426 24.3523C42.1563 24.5379 42.3002 24.6629 42.4744 24.7273C42.6525 24.7879 42.84 24.8182 43.0369 24.8182C43.1847 24.8182 43.3059 24.8106 43.4006 24.7955C43.4953 24.7765 43.571 24.7614 43.6278 24.75L43.9006 25.9545C43.8097 25.9886 43.6828 26.0227 43.5199 26.0568C43.357 26.0947 43.1506 26.1136 42.9006 26.1136C42.5218 26.1136 42.1506 26.0322 41.7869 25.8693C41.4271 25.7064 41.1278 25.4583 40.8892 25.125C40.6544 24.7917 40.5369 24.3712 40.5369 23.8636V15.1818Z" fill="#1F1F1F"/>
<rect x="0.5" y="0.5" width="292" height="39" rx="3.5" stroke="#545454"/>
</svg>
