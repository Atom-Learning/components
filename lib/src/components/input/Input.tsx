import * as React from 'react'

import { styled } from '~/stitches'
import { disabledStyle } from '~/utilities'
import { Override } from '~/utilities/types'

const StyledInput = styled('input', {
  appearance: 'none',
  border: '1px solid $tonal400',
  borderRadius: '$0',
  boxShadow: 'none', // prevent default iOS default styling
  boxSizing: 'border-box',
  color: '$tonal600',
  cursor: 'text',
  display: 'block',
  fontFamily: '$body',
  px: '$3',
  transition: 'all 100ms ease-out',
  width: '100%',
  '&:focus': {
    borderColor: '$primary800',
    outline: 'none'
  },
  '&[disabled]': disabledStyle,
  '&::placeholder': {
    color: '$tonal300',
    opacity: 1
  },
  variants: {
    size: {
      sm: {
        height: '$3',
        fontSize: '$sm',
        lineHeight: 1.7
      },
      md: {
        height: '$4',
        fontSize: '$md',
        lineHeight: 2
      },
      lg: {
        height: '$5',
        fontSize: '$md',
        lineHeight: 2
      },
      xl: {
        height: '$6',
        fontSize: '$lg',
        lineHeight: 2
      }
    },
    state: {
      error: {
        border: '1px solid $danger'
      }
    }
  }
})

// override default 'type' property to prevent Input from being used to render
// checkboxes, radios etc â€” we have dedicated components for them
export type InputProps = Override<
  React.ComponentProps<typeof StyledInput>,
  {
    name: string
    as?: never
    type?: 'text' | 'number' | 'email' | 'password' | 'tel' | 'url' | 'search'
  }
>

export const Input: React.ForwardRefExoticComponent<InputProps> =
  React.forwardRef(({ type = 'text', size = 'md', ...rest }, ref) => {
    if (type === 'number') {
      return (
        <StyledInput
          type="text"
          inputMode="numeric"
          pattern="[0-9]*"
          size={size}
          {...rest}
          ref={ref}
        />
      )
    }

    return <StyledInput type={type} size={size} {...rest} ref={ref} />
  })

Input.displayName = 'Input'
