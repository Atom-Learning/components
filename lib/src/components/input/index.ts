export type { InputProps } from './Input'
export { Input, InputBackground, InputText } from './Input'
