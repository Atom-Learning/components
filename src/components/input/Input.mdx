---
title: Input
component: Input
description: Input is the lowest-level input component for text and number types
category: Forms
---

Inputs should be accompanied by labels, so rather than using `Input` directly in a UI,
itâ€™s normally best to use a field component, which combines an `Input` with a `Label` and
displays validation errors. If none of the existing field components suit your needs,
it might be worth adding a new one.

`Input` accepts a subset of the `types` that an HTML `input` element because we have more specialised
components for the others (e.g. `Checkbox`, `Radio`).

```tsx live
<Input placeholder="Placeholder text" css={{ width: 300 }} />
```

## Accessibility

While `Input` accepts `"number"` for its `type` prop, this does not render `<input type="number"/>`. Instead,
it renders `<input type="text" inputmode="numeric" pattern=[0-9]*/>` to tell browsers to show a numeric keyboard
while avoiding the [UX and accessibility problems associated with `type="number"`](https://technology.blog.gov.uk/2020/02/24/why-the-gov-uk-design-system-team-changed-the-input-type-for-numbers/).
