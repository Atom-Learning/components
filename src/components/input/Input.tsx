import * as React from 'react'

import { styled } from '~/stitches'
import { Override } from '~/utilities/types'

const StyledInput = styled('input', {
  appearance: 'none',
  border: '1px solid $tonal500',
  borderRadius: '$0',
  boxShadow: 'none', // prevent default iOS default styling
  boxSizing: 'border-box',
  color: '$tonal900',
  cursor: 'text',
  display: 'block',
  fontFamily: '$sans',
  fontSize: '$md', // prevent iOS zooming on focus
  height: '$4',
  width: '100%',
  p: '$3',
  transition: 'all 100ms ease-out',
  '&:focus': {
    borderColor: '$primary900',
    boxShadow: 'inset 0 0 0 1px $colors$primary900',
    outline: 'none'
  },
  '&[disabled]': {
    backgroundColor: '$tonal300',
    color: '$tonal700',
    cursor: 'not-allowed'
  },
  variants: {
    state: {
      error: {
        border: '1px solid $danger'
      }
    }
  }
})

// override default 'type' property to prevent Input from being used to render
// checkboxes, radios etc â€” we will have dedicated components for them
export type InputProps = Override<
  React.ComponentProps<typeof StyledInput>,
  {
    as?: never
    type?: 'text' | 'number' | 'email' | 'password' | 'tel' | 'url'
  }
>

export const Input: React.FC<InputProps> = React.forwardRef(
  ({ type = 'text', ...rest }, ref) => {
    if (type === 'number') {
      return (
        <StyledInput
          type="text"
          inputMode="numeric"
          pattern="[0-9]*"
          ref={ref}
          {...rest}
        />
      )
    }

    return <StyledInput type={type} {...rest} ref={ref} />
  }
)

Input.displayName = 'Input'
