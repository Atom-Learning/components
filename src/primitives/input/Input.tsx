import { StitchesProps } from '@stitches/react'
import * as React from 'react'

import { styled } from '../../stitches'

const BaseInput = styled('input', {
  display: 'block',
  p: '$3',
  appearance: 'none',
  width: '100%',
  height: '50px',
  borderRadius: '$1',
  border: '1px solid $tonal500',
  boxShadow: 'none', // necessary to remove default iOS default styling
  fontSize: 'md', // necessary to prevent iOS zooming on focus
  fontFamily: 'sans',
  color: '$tonal900',
  transition: 'all 75ms ease-out',
  boxSizing: 'border-box',
  ':focus': {
    boxShadow: 'inset 0 0 0 1px $primary900',
    borderColor: '$primary900',
    outline: 'none'
  }
})

type InputProps = StitchesProps<typeof BaseInput>

export const Input = ({ type, ...props }: InputProps): React.ReactElement => {
  if (type === 'number')
    return (
      <BaseInput type="text" inputMode="numeric" pattern="[0-9]*" {...props} />
    )

  return <BaseInput type={type} {...props} />
}
