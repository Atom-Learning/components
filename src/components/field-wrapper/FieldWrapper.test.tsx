import { render, screen } from '@testing-library/react'
import userEvent from '@testing-library/user-event'
import { axe } from 'jest-axe'
import * as React from 'react'

import { Input } from '../input'
import { FieldWrapper } from '.'

const ExampleField = () => (
  <FieldWrapper label="Example Field" fieldId="example" error="Example error">
    <Input name="example" id="example" />
  </FieldWrapper>
)

describe('FieldWrapper component', () => {
  it('renders', async () => {
    const { container } = render(<ExampleField />)

    expect(container).toMatchSnapshot()
  })

  it('has no programmatically detectable a11y issues', async () => {
    const { container } = render(<ExampleField />)

    expect(await axe(container)).toHaveNoViolations()
  })

  it('links the label to an input', async () => {
    render(<ExampleField />)

    userEvent.click(await screen.findByText('Example Field'))

    expect(await screen.findByRole('textbox')).toHaveFocus()
  })

  it('links renders provided error', async () => {
    render(<ExampleField />)

    expect(await screen.findByText('Example error')).toBeVisible()
  })
})
