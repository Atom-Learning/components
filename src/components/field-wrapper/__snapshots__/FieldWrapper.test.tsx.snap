// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`FieldWrapper component renders 1`] = `
.sx47aq8 {
  -webkit-appearance: none;
  appearance: none;
  border: 1px solid var(--sx-colors-tonal500);
  border-radius: var(--sx-radii-0);
  box-shadow: none;
  box-sizing: border-box;
  color: var(--sx-colors-tonal900);
  cursor: text;
  display: block;
  font-family: var(--sx-fonts-sans);
  font-size: var(--sx-fontSizes-md);
  height: var(--sx-sizes-4);
  width: 100%;
  padding: var(--sx-space-3);
  transition: all 100ms ease-out;
}

.sx47aq8:focus {
  border-color: var(--sx-colors-primary900);
  box-shadow: inset 0 0 0 1px var(--sx-colors-primary900);
  outline: none;
}

.sx47aq8[disabled] {
  background-color: var(--sx-colors-tonal300);
  color: var(--sx-colors-tonal700);
  cursor: not-allowed;
}

.sx2cs53 {
  display: flex;
}

.sx2cs53-re4mk {
  justify-content: space-between;
  align-items: center;
  margin-bottom: var(--sx-space-2);
}

.sxfey5o {
  color: var(--sx-colors-tonal900);
  font-family: var(--sx-fonts-sans);
  font-weight: 400;
  margin: 0;
}

.sxfey5oxzxpn--size-sm {
  font-size: var(--sx-fontSizes-sm);
  line-height: 1.69;
}

.sxfey5oxzxpn--size-sm::before {
  content: '';
  margin-bottom: -0.477em;
  display: table;
}

.sxfey5oxzxpn--size-sm::after {
  content: '';
  margin-top: -0.477em;
  display: table;
}

.sxfey5o-m6cy5 {
  color: var(--sx-colors-danger);
  margin-top: var(--sx-space-1);
}

.sxdiu5c {
  color: var(--sx-colors-secondary300);
  display: block;
  font-family: var(--sx-fonts-sans);
  font-weight: 500;
  margin: 0;
}

.sxdiu5cjqqvr--size-md {
  font-size: var(--sx-fontSizes-md);
  line-height: 1.625;
}

.sxdiu5cjqqvr--size-md::before {
  content: '';
  margin-bottom: -0.4489em;
  display: table;
}

.sxdiu5cjqqvr--size-md::after {
  content: '';
  margin-top: -0.4489em;
  display: table;
}

.sxfxyev {
  background: unset;
  border: unset;
  padding: unset;
  color: var(--sx-colors-primary500);
  cursor: pointer;
  font-family: var(--sx-fonts-sans);
  text-decoration: none;
}

.sxfxyev:focus,
.sxfxyev:hover {
  color: var(--sx-colors-primary900);
  text-decoration: underline;
}

.sxfxyev:active {
  color: var(--sx-colors-primary500);
}

.sxfey5o > .sxfxyev,
.sx2g5mn > .sxfxyev,
.sx03kze > .sxfxyev {
  font-size: 100%;
  line-height: 1;
}

.sxfey5o > .sxfxyev::before,
.sxfey5o > .sxfxyev::after,
.sx2g5mn > .sxfxyev::before,
.sx2g5mn > .sxfxyev::after,
.sx03kze > .sxfxyev::before,
.sx03kze > .sxfxyev::after {
  content: none;
}

.sxfxyevxzxpn--size-sm {
  font-size: var(--sx-fontSizes-sm);
  line-height: 1.69;
}

.sxfxyevxzxpn--size-sm::before {
  content: '';
  margin-bottom: -0.477em;
  display: table;
}

.sxfxyevxzxpn--size-sm::after {
  content: '';
  margin-top: -0.477em;
  display: table;
}

<div>
  <div
    class="sx03kze"
  >
    <div
      class="sx2cs53 sx2cs53-re4mk"
    >
      <label
        class="sxdiu5c sxdiu5cjqqvr--size-md"
        for="example"
      >
        Example Field
      </label>
      <a
        class="sxfxyev sxfxyevxzxpn--size-sm"
        href="https://example.com"
      >
        Example prompt
      </a>
    </div>
    <input
      class="sx47aq8"
      id="example"
      name="example"
      type="text"
    />
    <p
      class="sxfey5o sxfey5oxzxpn--size-sm sxfey5o-m6cy5"
    >
      Example error
    </p>
  </div>
</div>
`;
